### ISSUE TITLE
Improve visual consistency and linking between Doxygen/Sphinx docs

### ISSUE THREAD

**Issue by bryanwweber** [Cantera/enhancements#115]
**Abstract**

The project [doxyrest](https://github.com/vovkos/doxyrest) provides an XML->reST compiler. This can convert Doxygen-generated XML documentation into the reST format that Sphinx understands. This would allow common theming of the C++ docs, rather than just sticking the Doxygen HTML into the website. It _may_ also allow linking via regular Docutils roles, so we don't need breathe/doxylink any longer.

**Motivation**

* What problem is it trying to solve?  
  Users and developers have noted that the distinct style of the C++ docs makes it seem like a different project and not connected to the rest of the docs. This would allow us to theme the C++ docs with the Sphinx engine that already generates the Python/Matlab docs.
* Who is affected by the change?
   All users of the website
* Why is this a good solution?
  I suppose another possible solution would be to write a Sphinx theme that fits the default Doxygen theme, or try to custom-theme the Doxygen docs. This seems a lot simpler, if it works.

**Comment by ischoegl** [comment]
:+1: ... this sounds quite promising! The only potential pitfall I see here is that the project doesn't have a lot of traffic yet and depends on one main developer (see [here](https://github.com/vovkos/doxyrest/graphs/contributors)). I'd still be inclined to test this as an 'experimental' feature.

**Comment by speth** [comment]
I'm confused by the comment regarding `breathe` -- we don't currently use that anywhere, do we? I had been considering it, in combination with [Exhale](https://github.com/svenevs/exhale) (the punny names are...intense) as an option to achieve this goal of rendering the C++ documentation with Sphinx.

One question that I have about `doxyrest` is whether it provides the sort of fully-automated API documentation generation that we currently get out of Doxygen, or whether it requires us to provide `.rst` files with entry points to the docs as we do now with the Python module (e.g. [`thermo.rst`](https://github.com/Cantera/cantera/blob/main/doc/sphinx/cython/thermo.rst)). This level is essentially the difference between `breathe` and the combination of `breathe` + `exhale`.

If the goal is just something with a bit more visual consistency, another alternative to consider would be to spend some time on the CSS template for the Doxygen docs. My guess is that's going to be a lot less work, and a lot less maintenance, than using any of these other packages.

**Comment by bryanwweber** [comment]
> I'm confused by the comment regarding `breathe` -- we don't currently use that anywhere, do we? I had been considering it, in combination with [Exhale](https://github.com/svenevs/exhale) (the punny names are...intense) as an option to achieve this goal of rendering the C++ documentation with Sphinx.

No, we don't use `breathe` but I had likewise been considering it.

> If the goal is just something with a bit more visual consistency

I think this is one goal, the other is to make linking easier, in conjunction with https://github.com/Cantera/cantera-website/issues/16. I'll change the title of the issue here

**Comment by speth** [comment]
One option for just improving the visual styling while continuing to use the Doxygen-generated HTML is to use https://github.com/jothepro/doxygen-awesome-css. With this package, you can get pages that look like [this](https://bluebrain.github.io/HighFive/class_high_five_1_1_data_set.html), with additional room for customization.

**Comment by ischoegl** [comment]
Fwiw, this is an interesting [SO post](https://stackoverflow.com/questions/70293037/how-to-render-doxygen-xml-output-into-markdown-instead-of-rest-using-breathe-or?noredirect=1&lq=1)

**Comment by ischoegl** [comment]
I played around with the various options for a small side-project of mine, and wanted to share some of my findings:

* [breathe](https://github.com/breathe-doc/breathe) is straight-forward to set up, but for output to work, some `*.rst` files need to be customized, as new Sphinx-styled documentation needs to be generated from doxygen's XML output. There is a 'kitchen-sink' approach (`doxygenindex` directive), alghough the `doxygengroup` directive is likely the most efficient approach; anything more fine-grained appears to be quite tedious. While the resulting output is consistent with the Sphinx layout, it isn't especially space-efficient when compared to doxygen. Obviously this can be addressed by optimizing css files and/or themes.
* [doxygen-awesome-css](https://github.com/jothepro/doxygen-awesome-css) is likewise easy to set up, as it is just an alternative layout where only a few flags in the `Doxyfile` need to be altered. There are quite a few custom user themes from the community posted [here](https://github.com/jothepro/doxygen-awesome-css/discussions/13)
* doxygen has limited Python support (although it doesn't appear to understand Cython), and Sphinx extensions are capable of handling C++, but neither documenting system appears to be particularly good outside of their main area.
* I don't believe there's a way to preserve collaboration diagrams, etc. information with `breathe` by itself; `exhale` may do this partially (?)
* I didn't test [exhale](https://github.com/svenevs/exhale) as it's supported by a single dev 😢; it would presumably automate the doxygen handling.
* *PS: I also didn't test `doxyrest` for similar reasons (single dev).*

After running through those tests, I believe that `doxygen` (especially with updated css styling and taking advantage of markdown support) may have an edge still. If you want to check out my tests, - they are rudimentary as I have no intent to make them permanent, - feel free to have a look at https://github.com/ischoegl/ctapp/tree/build-docs

PS: regardless of the verdict, Cantera's doxygen `@defgroup`'s need to be revised, see Cantera/cantera#1535.

**Comment by ischoegl** [comment]
Fwiw, a [blog post](https://itnext.io/daily-bit-e-of-c-modern-documentation-tools-9b96ba283732) relevant to this topic (doxygen/breathe/exhale + GH Actions) … with [rendered result](https://simontoth.eu/modern-documentation/)

**Comment by ischoegl** [comment]
Presumably resolved by #179/Cantera/cantera#1546 and Cantera/cantera#1573; the remainder is superseded by #182.

**Comment by ischoegl** [comment]
As it just popped up on my GitHub feed, there appears to be another project of interest: [doxysphinx](https://github.com/boschglobal/doxysphinx) ... as it apparently preserves some important doxygen features, it's definitely an interesting approach (@speth may already be looking into this).

**Comment by speth** [comment]
Yes, I ran across this last night, mainly while trying to figure out if there was a way to link into the Sphinx docs from Doxygen (which I haven't found yet). I wonder if this appeared for you because I starred it?

It does look interesting - this preserves most of the Doxygen HTML unmodified, so you get the class documentation layout that I think @ischoegl was specifically interested in preserving, but the outer navigational structure provided by Sphinx.

**Comment by ischoegl** [comment]
Yes, that's how I found out 😄 ... I found it interesting enough to follow up here. Fwiw, it's definitely not an urgent issue, but it does look like there *may* be a package that does exactly what we want after all. 

> this preserves most of the Doxygen HTML unmodified, so you get the class documentation layout that I think @ischoegl was specifically interested in preserving, but the outer navigational structure provided by Sphinx.

This is correct. 

What's interesting is that it may provide for a means for bi-directional content integration (sphinx <-> doxygen) - which is what you already pointed out. At the same time, we don't know whether it plays nice with doxylink (which is mostly unidirectional) or whether it simply supersedes that ... as said, I don't view it as urgent.
